### PR TITLE
MULE-11588: Enable bidirectional XML generation/parsing

### PR DESCRIPTION
**Pull request by alegmarra**

### CODE REVIEW

File: src/main/java/org/mule/runtime/api/app/declaration/EnrichableElementDeclaration.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import org.mule.runtime.api.meta.model.parameter.ParameterizedModel;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * A programmatic descriptor of a {@link ParameterizedModel} configuration.
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/app/declaration/EnrichableElementDeclaration.java]: doesn't match the class name

File: src/main/java/org/mule/runtime/api/app/declaration/FlowElementDeclaration.java [hunk 2]
```diff
@@ -43,77 +40,23 @@ public FlowElementDeclaration addComponent(ComponentElementDeclaration declarati
...
-    this.name = name;
-  }
-
   @Override
   public boolean equals(Object o) {
     if (this == o) {
       return true;
     }
-    if (!(o instanceof FlowElementDeclaration)) {
+    if (!(o instanceof FlowElementDeclaration) || !super.equals(o)) {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/app/declaration/FlowElementDeclaration.java]: that super.equals() seems like a hack which is hiding something else

File: src/main/java/org/mule/runtime/api/app/declaration/ParameterizedElementDeclaration.java [hunk 3]
```diff
@@ -16,7 +16,7 @@
  *
  * @since 1.0
  */
-public class ParameterizedElementDeclaration extends ElementDeclaration {
+public class ParameterizedElementDeclaration extends EnrichableElementDeclaration {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/app/declaration/ParameterizedElementDeclaration.java]: The fact that ParameterizedElementDeclaration extends EnrichableElementDeclaration confirms that the latter's jdoc is wrong

File: src/main/java/org/mule/runtime/api/app/declaration/RouterElementDeclaration.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  public void addRoute(RouteElementDeclaration route) {
+    this.routes.add(route);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof RouterElementDeclaration) || !super.equals(o)) {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/app/declaration/RouterElementDeclaration.java]: again, the super.equals doesn't seem right

File: src/main/java/org/mule/runtime/api/app/declaration/fluent/EnrichableElementDeclarer.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.app.declaration.fluent;
+
+import org.mule.runtime.api.app.declaration.ElementDeclaration;
+import org.mule.runtime.api.app.declaration.EnrichableElementDeclaration;
+import org.mule.runtime.api.app.declaration.ParameterElementDeclaration;
+
+/**
+ * //TODO
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/app/declaration/fluent/EnrichableElementDeclarer.java]: do it

File: src/main/java/org/mule/runtime/api/app/declaration/fluent/ParameterizedBuilder.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.api.app.declaration.fluent;
+
+/**
+ * //TODO
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/app/declaration/fluent/ParameterizedBuilder.java]: do it

File: src/main/java/org/mule/runtime/api/app/declaration/fluent/EnrichableElementDeclarer.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+ */
+class EnrichableElementDeclarer<D extends EnrichableElementDeclarer, T extends EnrichableElementDeclaration>
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/app/declaration/fluent/EnrichableElementDeclarer.java]: I'm going to charge you royalties for these declarers ;)

File: src/main/java/org/mule/runtime/api/meta/model/util/ComponentModelVisitor.java [hunk 8]
```diff
@@ -24,26 +24,26 @@
    *
    * @param operationModel the visited object
    */
-  void visit(OperationModel operationModel);
+  default void visit(OperationModel operationModel) {}
```
Comment (marianogonzalez) [inline src/main/java/org/mule/runtime/api/meta/model/util/ComponentModelVisitor.java]: not convenient to do this until Scope and Route model stop extending OperationModel

File: src/main/java/org/mule/runtime/api/app/declaration/RouterElementDeclaration.java (hunk 4 above)
Comment (alegmarra) [inline src/main/java/org/mule/runtime/api/app/declaration/RouterElementDeclaration.java]: I don't get why this is bad. It makes sense to delegate the equality check to the parent that already implemented the comparison for it's own fields.
I've looked for examples and the equals method suggested intellij adds this by default, equal methods used by java.util classes use this `super.equals()` delegation (check java.util.GregorianCalendar#equals), and even apache's EqualsBuilder provides an `appendSuper` method  for the builder.

File: src/main/java/org/mule/runtime/api/app/declaration/ParameterizedElementDeclaration.java (hunk 3 above)
Comment (alegmarra) [inline src/main/java/org/mule/runtime/api/app/declaration/ParameterizedElementDeclaration.java]: fixed

**Pull request merged**